### PR TITLE
Provide an EnumCoproductHint for sealed traits of case objects

### LINKED ISSUE #208: Provide a CoproductHint to serialize sealed families of case objects
One very common use case I have with pureconfig is to serialize enums (encoded as sealed families of case objects) as simple strings with the names of the options. I think that it is a case common enough for us to provide a `CoproductHint`.

### PR DESCRIPTION
**Pull request by ruippeixotog**
Closes #208.

### CODE REVIEW

File: core/src/main/scala/pureconfig/CoproductHint.scala
Comment (melrief) [inline core/src/main/scala/pureconfig/CoproductHint.scala:108]: The reason why this is protected is to give devs a hint at the fact that they can subclass this class and change how the `fieldValue` is calculated from a `name`?

File: core/src/main/scala/pureconfig/CoproductHint.scala
Comment (melrief) [inline core/src/main/scala/pureconfig/CoproductHint.scala]: I would add that the `string` used is the object name, e.g. `object Foo` will be read as `foo` with this. It's not super clear how objects are written and read as strings.

File: core/src/test/scala/pureconfig/CoproductHintSuite.scala
Comment (melrief) [inline core/src/test/scala/pureconfig/CoproductHintSuite.scala:54]: Nice!

**Review by melrief — APPROVED** [review]
LGTM, good to have this!

File: core/src/main/scala/pureconfig/CoproductHint.scala
Comment (ruippeixotog) [reply core/src/main/scala/pureconfig/CoproductHint.scala:108]: Yes, it follows the same structure as `FieldCoproductHint`, where the translation between class names and config values is customizable.

Just as we did with `ProductHint` (following [the discussion here](https://github.com/pureconfig/pureconfig/pull/90#discussion_r97449022)), we could provide a constructor for both of these hints that receives a `ConfigFieldMapping`. I'd rather keep a consistent interface between hints here and leave a possible refactoring or new constructor for another PR.

File: core/src/main/scala/pureconfig/CoproductHint.scala
Comment (ruippeixotog) [reply core/src/main/scala/pureconfig/CoproductHint.scala]: I agree, I clarified it in 1fab2b3.

**Comment by melrief** [comment]
LGTM, waiting for a second review

**Review by melrief — APPROVED** [review]

**Comment by melrief** [comment]
coveralls seems unhappy with the test coverage. It's not a blocker for the PR but maybe we can make a few more tests to cover the new code?

**Comment by ruippeixotog** [comment]
If you see the list of changes in coverage, most of them are not related to this PR (or at least I cannot see what might have affected them), which is a bit strange. Maybe some of our tests are non-deterministic?

Regarding the lines related to this PR that are uncovered, so you want me to add tests for things like the description of the `ConfigReaderFailure`? I can test the exception message with a regex, if you think it's worth it.

**Comment by melrief** [comment]
No @ruippeixotog, it's not worth it. I'm trying to understand what's going on.

**Review by jcazevedo — APPROVED** [review]

**Pull request merged**
